rcl_lifecycle/src/com_interface.c

```diff
@@ -0,0 +1,203 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include <stdio.h>
+#include <string.h>
+
+#include "rosidl_generator_c/message_type_support.h"
+#include "rosidl_generator_c/string_functions.h"
+
+#include "lifecycle_msgs/msg/transition_event.h"
+#include "lifecycle_msgs/srv/get_state.h"
+#include "lifecycle_msgs/srv/change_state.h"
+
+#include "rcl/error_handling.h"
+
+#include "rcl_lifecycle/data_types.h"
+
+#include "com_interface.hxx"
+
+static lifecycle_msgs__msg__TransitionEvent msg;
+
+bool concatenate(const char ** prefix, const char ** suffix, char ** result)
+{
+  size_t prefix_size = strlen(*prefix);
+  size_t suffix_size = strlen(*suffix);
+  if ((prefix_size + suffix_size) >= 255) {
+    return false;
+  }
+  *result = malloc((prefix_size + suffix_size) * sizeof(char));
+  memcpy(*result, *prefix, prefix_size);
+  memcpy(*result + prefix_size, *suffix, suffix_size + 1);
+  return true;
+}
+
+rcl_lifecycle_com_interface_t
+rcl_lifecycle_get_zero_initialized_com_interface()
+{
+  rcl_lifecycle_com_interface_t com_interface;
+  com_interface.state_publisher = rcl_get_zero_initialized_publisher();
+  com_interface.srv_get_state = rcl_get_zero_initialized_service();
+  com_interface.srv_change_state = rcl_get_zero_initialized_service();
+
+  return com_interface;
+}
+
+rcl_ret_t
+rcl_lifecycle_com_interface_init(rcl_lifecycle_com_interface_t * com_interface,
+  rcl_node_t * node_handle,
+  const rosidl_message_type_support_t * ts_pub_notify,
+  const rosidl_service_type_support_t * ts_srv_get_state,
+  const rosidl_service_type_support_t * ts_srv_change_state)
+{
+  if (!ts_pub_notify) {
+    RCL_SET_ERROR_MSG("ts_pub_notify is NULL");
+    return RCL_RET_ERROR;
+  }
+  if (!ts_srv_get_state) {
+    RCL_SET_ERROR_MSG("ts_srv_get_state is NULL");
+    return RCL_RET_ERROR;
+  }
+  if (!ts_srv_change_state) {
+    RCL_SET_ERROR_MSG("ts_srv_change_state is NULL");
+    return RCL_RET_ERROR;
+  }
+
+
+  const char * node_name = rcl_node_get_name(node_handle);
+
+  {  // initialize publisher
+     // Build topic, topic suffix hardcoded for now
+     // and limited in length of 255
```

Comment:
The three space indentation looks weird. Maybe don't put the first comment in line 85 but into a new line. Than it can have consistent indentation with the block.

Same below.